Feature: Revamping the whole project

vp-en opened the pull request:
Hey there. I did a restructure/revamp of the entire `red2lem` project. Here are some of the highlights:


## Changes
1. **Environment Variable Usage**: The new code uses environment variables (`os.environ`) for configuration, while the old code was relying on separate credential files and variables. This allows for more flexible deployment, especially in a containerized environment like Docker.

2. **Detection of Running Environment**: The new code has a condition to detect if it's running inside a container, potentially allowing for different behavior or settings when running within Docker.

3. **Enhanced Media Handling and Imgur Integration**: In the new code, there is logic for uploading videos to Imgur if an Imgur client ID/secret ID is provided. This means that videos can be re-uploaded to a different hosting service, whereas the old code directly used Reddit's media URLs.

4. **Fallback to JSON When API Calls Fail**: The new code introduces a function (`get_submission_from_json`) to fetch Reddit submission data from a JSON endpoint if the regular API call fails. This can enhance the reliability of the script.

5. **Original Reddit Post Link**: The new code adds a link to the original Reddit post in the Lemmy post, using the `np.reddit.com` domain.

6. **Configurable Sleep Time**: In the new code, the sleep time between checking for new posts is configurable via an environment variable, with a default value of 60 seconds.

7. **Refactoring and Reorganization**: Functions have been restructured, and the code has been organized into a `main` function, improving readability and maintainability.

8. **Logging and Error Handling**: There's more detailed logging in the new code, and more error handling to gracefully deal with exceptions. For example, it checks if a Lemmy post was created successfully and posts the link to it.

In summary, the new code has been considerably changed, introducing new features, and offering more customization options for the user.

The code could probably be better or whatever, but it works pretty well right now. 


## Testing this PR
If you want to try the bot as it is, you can use the following `docker compose` example:

```yaml
version: "3.9"
services:
  red2lem:
    build: https://github.com/vp-en/red2lem.git#feat-revampProject
    container_name: red2lem
    environment:
      reddit_client_id: YOUR_REDDIT_CLIENT_ID
      reddit_client_secret: YOUR_REDDIT_CLIENT_SECRET
      lemmy_url: YOUR_LEMMY_URL
      lemmy_username: YOUR_LEMMY_USERNAME
      lemmy_password: YOUR_LEMMY_PASSWORD
      subreddit_name: YOUR_SUBREDDIT_NAME
      community_name: YOUR_COMMUNITY_NAME
      imgur_client_id: YOUR_IMGUR_CLIENT_ID # Optional
      imgur_client_secret: YOUR_IMGUR_CLIENT_SECRET # Optional
      copy_automoderator_posts: false # Optional, default is False
      post_check_interval: 60 # Optional, default is 60
      title_template: "[X-Posted from Reddit - /u/{author_name}] - {submission_title}" # Optional
    restart: unless-stopped
```

Commit:
Revamp

Reworking most of the project, it works pretty good now

Commit:
Adding `title_template`

It'll keep the default as `[X-Posted from Reddit - /u/{author_name}] - {submission_title}`, but you can now use a custom title with that ENV.

The {author_name} and {submission_title} can be used in `title_template`

ghost commented:
Now I gotta be honest, I haven't tested this in a non-Docker environment, but it works very well using Docker.

**Edit:**

Quick note.. The ENV `community_name` must include the instance url as well, e.g. `bot_testing@lemm.ee`

Commit:
Better logging

This should include better logging, and a few miscellaneous fixes

Fix sleep time
